src/utils/atomFamily.ts

```diff
@@ -61,45 +60,52 @@ export function atomFamily<Param, Value, Update extends never = never>(
 export function atomFamily<Param, Value, Update>(
   initializeRead: (param: Param) => any,
   initializeWrite?: null | ((param: Param) => any),
-  areEqual: (a: Param, b: Param) => boolean = Object.is
+  areEqual?: (a: Param, b: Param) => boolean
 ) {
   type AtomType = WritableAtom<Value, Update>
   type CreatedAt = number // in milliseconds
   let shouldRemove: ShouldRemove<Param> | null = null
-  const atoms: [Param, AtomType, CreatedAt][] = []
+  const atoms: Map<Param, [AtomType, CreatedAt]> = new Map()
   const createAtom = (param: Param) => {
-    const index = atoms.findIndex((x) => areEqual(x[0], param))
-    if (index >= 0) {
-      const item = atoms[index]
-      if (shouldRemove && shouldRemove(item[2], item[0])) {
-        atoms.splice(index, 1)
+    let item: [AtomType, CreatedAt] | undefined
+    if (areEqual === undefined) {
+      item = atoms.get(param)
+    } else {
+      // Custom comparator, iterate over all elements
+      for (let [key, value] of atoms) {
+        if (areEqual(key, param)) {
+          item = value
+          break
+        }
+      }
+    }
+
+    if (item !== undefined) {
+      if (shouldRemove?.(item[1], param)) {
+        atoms.delete(param)
       } else {
-        return item[1]
+        return item[0]
       }
     }
+
     const newAtom = atom(
       initializeRead(param),
       initializeWrite && initializeWrite(param)
     ) as AtomType
-    atoms.unshift([param, newAtom, Date.now()])
+    atoms.set(param, [newAtom, Date.now()])
     return newAtom
   }
+
   createAtom.remove = (p: Param) => {
-    const index = atoms.findIndex((x) => x[0] === p)
```

Comment:
Just noticed my original implementation doesn't take care of `areEqual`.
So, this is not a regression in this PR, but we should use areEqual here too to be consistent.

Reply:
#177 follow up issue.